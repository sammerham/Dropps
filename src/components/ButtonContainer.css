.buttonContainer-btn{
    height:50px;
    font-size: 20px;
    margin: 20px;
    background-color: lightyellow;
    border-color: gray;
}


.buttonContainer-div {
  margin-left:20px;
  font-size:18px;
}

.buttonContainer-span{
  color:red;
}