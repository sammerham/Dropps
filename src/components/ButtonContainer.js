import React, {useState, useEffect, useRef} from 'react'
import axios from 'axios';
import './ButtonContainer.css'


/** ButtonContainer
 * 
 * Props: 
 *  - none
 * State:
 *  - counter : Number
 *  - origin : String
 *  - loading (boolean)
 *  - clicked (boolean)
 * 
 * App -> ButtonContainer 
 */

const ButtonContainer = () => {
  const [counter, setCounter] = useState(0);
  const [origin, setOrigin] = useState('');
  const [loading, setLoading] = useState(false);
  const [clicked, setClicked] = useState(false);
  const abortController = useRef(null);


  //*effect to add 1 to counter and update local storage with origin when clicked changed*/

  useEffect(() => {
    setCounter(counter => counter + 1);
    const savedOrigin = localStorage.getItem("orig")
    setOrigin(savedOrigin)
  }, [clicked]);
//******************************************************************************************************** */
  
  //* effect to call api runs when clicked variable changes 
  useEffect(() => {
    const fetchData = async () => {
      //! if button is clicked makes the request
    if (clicked) 
    try {
      setLoading(true)
      // abour logic
      abortController.current = new AbortController();
      let signal = abortController.current.signal;
      // const res = await axios.get('https://httpbin.org/status/400', { signal });
      const res = await axios.get('https://httpbin.org/delay/4', { signal });

      //! saved origin in local storage when new origin comes from api
      localStorage.setItem("orig", res.data.origin);
      setOrigin(res.data.origin);
      setLoading(false);
      setClicked(false);
    } catch (err) {
      setLoading(false);
      setClicked(false);
      console.warn(err);
      };
    };
    fetchData();
    //* this func will run when the button comp unmounts (clean up function)
    return () => {
      abortController.current && abortController.current.abort()
    };
  },[clicked])


// fn to handle Cancel button click.
  const handleCacel = () => { 
    setLoading(false)
    setClicked(false)
    abortController.current && abortController.current.abort();
  }

  // fn to handle Click Me button click.
  const handleClick = () => setClicked(true);


  return (
    <div>
      <button
        onClick={handleClick}
        disabled={loading}
        className="buttonContainer-btn"
      >
        {loading ? "Click Me (loading)" :"Click Me" }
      </button>
      <button
        disabled={!loading}
        onClick={handleCacel}
        className="buttonContainer-btn"
      >
        Cancel
      </button>
      <br />
      <div className="buttonContainer-div">
        Renders: <span className='buttonContainer-span'>&nbsp;{counter}</span>
        <br />
        <br />
        Origin:<span className='buttonContainer-span'>&nbsp;{origin}</span>
      </div> 
    </div>
  )
}

export default ButtonContainer;