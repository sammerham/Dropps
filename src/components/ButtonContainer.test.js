import { render, screen, fireEvent } from '@testing-library/react';
import ButtonContainer from './ButtonContainer'


test('It renders without crashing', () => {
  render(<ButtonContainer />)
});

test('It matches snapshot', () => {
  const { container } = render(<ButtonContainer />);
  expect(container).toMatchSnapshot();
});


test('It should show disabled cancel button', () => {
  render(<ButtonContainer />);
  let cancelButton = screen.getByText('Cancel');
  expect(cancelButton).toBeInTheDocument();
  expect(cancelButton).toBeDisabled()
});


test('It should show enabled Click Me button', () => {
  render(<ButtonContainer />);
  let clickMeButton = screen.getByText('Click Me');
  expect(clickMeButton).toBeInTheDocument();
  expect(clickMeButton).toBeEnabled()
});

test('It should show cancel btn enabled and clicked me disabled when Clicked me is clicked', () => {
  render(<ButtonContainer />);
  let cancelButton = screen.getByText('Cancel');
  let clickMeButton = screen.getByText('Click Me');
  fireEvent.click(clickMeButton);
  expect(cancelButton).toBeEnabled()
  expect(clickMeButton).toBeDisabled()
});


test('It should show loading and be disabled when Clicked me is clicked', () => {
  render(<ButtonContainer />);
  let clickMeButton = screen.getByText('Click Me');
  fireEvent.click(clickMeButton);
  expect(clickMeButton).toHaveTextContent("Click Me (loading)")
  expect(clickMeButton).toBeDisabled()
});
