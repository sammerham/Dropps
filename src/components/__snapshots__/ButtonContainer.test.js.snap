// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It matches snapshot 1`] = `
<div>
  <div>
    <button
      class="buttonContainer-btn"
    >
      Click Me
    </button>
    <button
      class="buttonContainer-btn"
      disabled=""
    >
      Cancel
    </button>
    <br />
    <div
      class="buttonContainer-div"
    >
      Renders: 
      <span
        class="buttonContainer-span"
      >
         
        1
      </span>
      <br />
      <br />
      Origin:
      <span
        class="buttonContainer-span"
      >
         
      </span>
    </div>
  </div>
</div>
`;
